python/hail/docs/style-guide.txt

```diff
@@ -1,32 +1,138 @@
+ - matplotlib
+ - pandas
+ - numpy
+ - seaborn
 
-::
+## Function Documentation Structure
+ - Description
+ - Examples
+ - Notes (if needed)
+ - Other subsections (if needed)
+ - Annotations (if needed)
+ - Parameter Specification
 
- >>> ...
+### Description
+ - Start with a short description
 
-Code examples should assume input data files are in data/.
+### Examples
+ - Create an examples section with this structure:
 
+    ```
+    **Examples**
+
+    Short description of example 1:
+
+    >>> python code example
+
+    Short description of example 2:
+
+    >>> python code example
+    ```
+
+ - The first example should be the most common use case.
+ - Try to keep example descriptions short and concise.
+
+### Additional Information
+ - This is not required. If needed use the heading `**Notes**`. Additional subsections can be added.
  - Put short python expressions or code snippets and Hail expressions
    in double-``.
 
- - Use :param:, :rtype: (if not None), :return:, and they should go at end.
+### Annotations
+ - For commands that create annotations, have "Annotations" section that lists the annotations.  Use a bulleted list with the following format:
+
+       ```
+       **Annotations**
 
+        - **annotation** (*Type*) -- description
+       ```
+
+### Parameter and Return Type Specification
+ - Use :param:, :rtype: (if not None), :return:.
+ - :return: gives a short description of what is being returned. Example: `A VariantDataset that has been annotated.`
+
+### General Style
  - links: `this is a description <with a url>`_.
+ - subsections: use `**`, we'll look into something better
+ - All function/command references should use a Sphinx directive to link ```:py:meth:`~hail.VariantDataset.vep````.
+
+## Code Examples
+ - All examples are automatically tested with the [Sphinx doctest extension](http://www.sphinx-doc.org/en/stable/ext/doctest.html) to make sure they run with no errors. The content of the result is not checked.
+ - All input files required must be placed in `python/hail/docs/data`. When referencing the files in the code example, the input directory is `data/` and the output directory is `output/`.
+ - Each command should start with `>>>`. If the command statement is on multiple lines, use `...` for each subsequent line.
+ - To skip execution of a command, see the `vep` example in `dataset.py`. Please try not to use this unless absolutely necessary.
+ - The HailContext `hc` is in scope and the following import statements have been run:
+
+    ```
+    from hail import *
+    from hail.representation import *
+    from hail.type import *
+    from hail.stats import *
+    ```
+
+ - Make sure you do not assign a result to the protected variable names `vds` in `dataset.py` and `kt1`, `kt2` in keytable.py.
+ - Variables not specified in the `.. testsetup::` directive in the module-level docstring will not be available in the scope for other functions. Try not to add global variables unless absolutely necessary.
```

Comment:
I was a bit confused by the reference to a ".. testsetup::" directive, in that I tried to seach for that text in file and couldn't find it anywhere. I see now that variables specified in module-level docstring are the only ones in scope for other functions, but I feel like if there isn't actually a tag of some kind that marks the testsetup directive, it's a little confusing.